stresstests/test/test_d2c_stress.js

```diff
@@ -0,0 +1,185 @@
+
+config.transportsToTest.forEach(transportToTest => {
+  config.deviceClientHelperMethodsToUse.forEach(deviceClientHelperMethodToUse => {
+    d2cStressTests(transportToTest, deviceClientHelperMethodToUse);
+  });
+});
+
+function d2cStressTests(transportToTest, deviceClientHelperMethodToUse) {
+  describe(`D2C stress tests on client created with ${deviceClientHelperMethodToUse} over transport ${transportToTest.name}`, function() {
+    let deviceClientHelper = null;
+    let eventHubHelper = null;
+    let stressMeasurementsRecorder = null;
+
+    before(function() {
+      // TODO: Remove test skipping when appropriate. 
+      if (['Amqp', 'AmqpWs'].includes(transportToTest.name)) {
+        debug(
+          'Tests over AMQP failing because of an issue with messageIds not '
+          + 'matching up on the service side. Skipping.'
+        );
+        this.skip();
+      }
+      if ([
+        'createDeviceClientX509SelfSigned',
+        'createDeviceClientX509CaSigned'
+      ].includes(deviceClientHelperMethodToUse)) {
+        debug(
+          'DeviceClientHelper has not implemented '
+          + `${deviceClientHelperMethodToUse}. Skipping.`
+        );
+        this.skip();
+      }
+
+      deviceClientHelper = new DeviceClientHelper();
+      eventHubHelper = new EventHubHelper();
+      stressMeasurementsRecorder = new StressMeasurementsRecorder();
+    });
+
+    beforeEach(async function() {
+      await eventHubHelper.open();
+      await deviceClientHelper[deviceClientHelperMethodToUse](transportToTest);
+      stressMeasurementsRecorder.start(
+        deviceClientHelper.client,
+        config.memorySamplingIntervalInMs
+      );
+    })
+
+    afterEach(async function() {
+      this.timeout(30 * 1000);
+      /*  done() should have been called within the test, but we call it here in
+          case the test failed and done() didn't get called. */
+      stressMeasurementsRecorder.done();
+      await Promise.all([
+        eventHubHelper.close(),
+        deviceClientHelper.disposeClient()
+      ]);
+    });
+
+    after(wtf.dump);
+
+    async function sendSingleMessage() {
+      const messageId = uuid.v4();
+      /* Function creates a string of random ASCII values between 32 and 126 */
+      const randomString = length => Array.from(
+        { length: Math.floor(length) },
+        () => String.fromCharCode(Math.floor(95 * Math.random() + 32))
+      ).join('');
+      /* 7 is the number of characters in {"":""} */
+      const payloadSize = Math.max(config.telemetryPayloadSizeInBytes, 7);
+      const keyLength = Math.ceil((payloadSize - 7) / 2);
+      const valLength = payloadSize - keyLength - 7;
+      const message = new Message(JSON.stringify({[randomString(keyLength)]: randomString(valLength)}));
+      message.messageId = messageId;
+      const deferred = eventHubHelper.awaitMessage(messageId);
+      stressMeasurementsRecorder.messageEnqueued();
+      let attempts = 0;
+      while (attempts < 5) {
```

Comment:
Doing retries here. The SDK is supposed to retry for us, at least for certain errors that are deemed retriable. I added this to reduce the flakiness of the tests. Should we keep this, or are we suppressing failures that are interesting to us by doing this?

Reply:
This depends on the API contract.  Right now, the Python contract doesn't have reliable retry, so the longhaul tests need to do it themselves.  Node, on the other hand, does have reliable retry, so this is unnecessary and probably hides failures that we'd want to see.

Reply:
Cool. I'll remove the retrying then.